### ISSUE TITLE
Add your site via .ac.uk verification (Future Development)

### ISSUE THREAD

**Issue by closebracket** [FreesideHull/hullblogs.com#1]
The present system for adding blogs to hullblogs requires manual json entry. If we're to make this process easier for students with limited technical knowledge and experience, we need to make this process as simple as possible. I would therefore like to recommend the following process and enchancement:

1. "Add blog button" added to top right of site. 
2. When clicking button, user is taken to form where they can provide the details required. 
3. User is also required to provide .ac.uk email address
4. User is required to verify the address 
5. If verified, the blog is added to the aggregator. If not verified, the information is disgarded after 14 days.

**Comment by sbrl** [comment]
Hey, great idea! Unfortunately, that's an absolutely *enormous* amount of effort, and is also inherently incompatible with the entire  design because I've used a static site generator. So to do that you would need to rewrite the entire website from scratch, which I'm not willing to do.

However, all is not lost. We could:

 - Use GitHub issue templates and have people open issues for their blog (I recommend this)
 - Use a Google Form

The other thing users must have here is a GitHub account, as I use it to get their avatar.

Why not Gravatar? Because then we would need to store their email addresses in the GitHub repository, and scrapers and spammers would have a field day.

**Comment by Epictek** [comment]
Why not use freeside accounts for managing blogs? Get to brand it with "powered by freeside" and the emails are already verified

**Comment by Epictek** [comment]
As for implementing that, could have a script/bot that pushes to the git repo using a template

**Comment by sbrl** [comment]
> Why not use freeside accounts for managing blogs? Get to brand it with "powered by freeside" and the emails are already verified

great idea, but that would also mean writing either:

 - A completely new website from scratch
 - A complete new and separate management interface

...both of which are an enormous amount of work @Epictek

**Comment by closebracket** [comment]
> 
> 
> > Why not use freeside accounts for managing blogs? Get to brand it with "powered by freeside" and the emails are already verified
> 
> great idea, but that would also mean writing either:
> 
>     * A completely new website from scratch
> 
>     * A complete new and separate management interface
> 
> 
> ...both of which are an enormous amount of work @Epictek

We could work on this as future development? Use the existing site now and then work on a v2?

**Comment by sbrl** [comment]
> We could work on this as future development? Use the existing site now and then work on a v2?

Possible, but the important thing here is that static site generators is inherently incompatible with what you're suggesting

You can't have a dynamic form on the website itself that auto-updates the `feeds.json` file. it doesn't work like that.

A static site generator does everything at build time, and spits out plain-old HTML/CSS/asset files that you can upload to a static web server.

So any solution is going to need to take that into account.

This is why I suggest for those who aren't confident with editing JSON to open an issue - we can easily create an issue template: https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository

then we could even have a direct link on the website to that.

The question does arise though how do we verify requests.

Finally, a GitHub account is required for fetching user avatars. I decided against Gravatar 'cause I don't want to store email addresses in this repo because of spam.

**Comment by closebracket** [comment]
> Possible, but the important thing here is that static site generators is inherently incompatible with what you're suggesting
I don't see how? Fundamentally what is being suggested is a form which sends information somewhere else to be dealt with, I'm quite sure a static site, or the infrastructure behind it can deal with that? It would be up to a separate part of the infrastructure to deal with the information recieved, request a verification and rebuild the site. 

> You can't have a dynamic form on the website itself that auto-updates the feeds.json file. it doesn't work like that.
That's not what is being suggested. The idea would be to create a backend script that appends the new blog to the ```feeds.json``` file if the user is successfully verified. I like @Epictek's suggestion of making users sign up to a Freeside account to do this, because we could then expand into other offerings as well - e.g. Linux tutorials with progress saved and a certifcate offered by say Canonical or something at the end (blue sky thinking, I know...) 

> A static site generator does everything at build time, and spits out plain-old HTML/CSS/asset files that you can upload to a static web server.
See [this list of static site form providers}(https://css-tricks.com/a-comparison-of-static-form-providers/).

> This is why I suggest for those who aren't confident with editing JSON to open an issue - we can easily create an issue template: https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository

That assumes people have a github account. Only one constant is certain: users will have a university email address. We don't know their department, technical proficency, accessibility issues, access to tech or any of those things. Hence why v2 of this needs to take that into account. It's OK right now, for the sake of speed, to push out the version you've created, and I agree that we shouldn't change the present version just to take into account this feature; but it does need to be added later.

> Finally, a GitHub account is required for fetching user avatars. I decided against Gravatar 'cause I don't want to store email addresses in this repo because of spam.

I think that's a separate issue all together to this one.

**Comment by sbrl** [comment]
> That's not what is being suggested. The idea would be to create a backend script that appends the new blog to the feeds.json file if the user is successfully verified.

That could work. But the Freeside account signup process isn't exactly battle tested... :P

Also you'd need integration with LDAP, which is a pain

Either way that would be a separate project in a separate repository to this one.

> That assumes people have a github account. Only one constant is certain: users will have a university email address. We don't know their department, technical proficency, accessibility issues, access to tech or any of those things.

Yeah, it does.

> Hence why v2 of this needs to take that into account. 

That's probably a good idea, but I'm unlikely to have the energy to develop something like that before I finish my PhD.

> I think that's a separate issue all together to this one.

Yes it is

**Comment by fisherthewol** [comment]
> 3. User is also required to provide .ac.uk email address

If it's HullBlogs, then if we go by email address, it's a restriction to @hull.ac.uk, surely?

Also, if we're on a static generator, then for fresh content we will need a process to do daily builds (IE Continuous Integration). So any changes to feeds.json will be picked up on the next build.
Hence, it seems relatively simple to write part of the CI chain that would ingress data, verify, then egress to feeds.json.

My proposal is similar to that of @Epictek and @closebracket. 
We can additional field in our IPA/LDAP systems, that allows a user to point at an RSS feed (I believe users can edit their own details? If not, we can change it so they can edit some of them).
Then, a script is run (seperately or as part of CI) that pulls these fields from LDAP ([python-ldap](https://www.python-ldap.org/en/python-ldap-3.3.0/) for python, [System.DirectoryServices.Protocols](https://docs.microsoft.com/en-us/dotnet/api/system.directoryservices.protocols) for C#), does some glue-work, and edits feeds.json.

I don't think this should be the only way a blog can be added, but I think it should be the primary way, so that it promotes people to sign up for freeside accounts - which is a process that needs work as well.

**Comment by sbrl** [comment]
> If it's HullBlogs, then if we go by email address, it's a restriction to @hull.ac.uk, surely?

Yep, that's pretty much it

> We can additional field in our IPA/LDAP systems, that allows a user to point at an RSS feed (I believe users can edit their own details? If not, we can change it so they can edit some of them).


That's a great idea. Perhaps we could update profiles.freeside.co.uk to allow you to specify a link to your blog feed perhaps? Then use some glue code to push that into `feeds.json`.

**Comment by fisherthewol** [comment]
Yep, that's what I was thinking - I need to get the VPN working again (winget updated it too far) and then I'll take a look.

**Comment by closebracket** [comment]
I'm happy with the suggestion made by @fisherthewol. Can we get it implemented before Sept so we can start promoting it at various places within the uni?
